exonum-java-binding/core/rust/ejb-app/src/node_builder.rs

```diff
@@ -17,16 +17,80 @@
 use exonum_btc_anchoring::ServiceFactory as BtcAnchoringServiceFactory;
 use exonum_configuration::ServiceFactory as ConfigurationServiceFactory;
 use exonum_time::TimeServiceFactory;
-use java_bindings::exonum::helpers::fabric::{self, ServiceFactory};
-use java_bindings::utils::{
-    load_enabled_services, BTC_ANCHORING_SERVICE, CONFIGURATION_SERVICE, EJB_SERVICE,
-    PATH_TO_SERVICES_TO_ENABLE, TIME_SERVICE,
+use java_bindings::{
+    exonum::helpers::fabric::{self, ServiceFactory},
+    utils::services::{
+        load_services_definition, EjbAppServices, BTC_ANCHORING_SERVICE, CONFIGURATION_SERVICE,
+        PATH_TO_SERVICES_DEFINITION, TIME_SERVICE,
+    },
+    JavaServiceFactoryAdapter,
 };
-use java_bindings::JavaServiceFactory;
 
-use std::collections::{HashMap, HashSet};
-use std::path::Path;
+use std::{
+    collections::{HashMap, HashSet},
+    path::Path,
+};
+
+/// Creates `NodeBuilder` using services configuration from `ejb_app_services.toml` located in the working directory.
+pub fn create() -> fabric::NodeBuilder {
+    let service_factories = prepare_service_factories(PATH_TO_SERVICES_DEFINITION);
+    let mut builder = fabric::NodeBuilder::new();
+    for service_factory in service_factories {
+        builder = builder.with_service(service_factory);
+    }
+    builder
+}
+
+// Prepares vector of `ServiceFactory` from services configuration file located at given path.
+// Panics in case of problems with reading/validating of service configuration file.
+fn prepare_service_factories<P: AsRef<Path>>(path: P) -> Vec<Box<dyn ServiceFactory>> {
+    // Read services definition from file.
+    let EjbAppServices {
+        system_services,
+        user_services,
+    } = load_services_definition(path).expect("Unable to load services definition");
+
+    // Make sure there is at least one user service defined.
+    if user_services.is_empty() {
+        panic!("At least one user service should be defined in the \"ejb_app_services.toml\" file");
+    }
 
+    // Check whether we have system services defined or insert the configuration service otherwise.
+    let system_services = system_services.unwrap_or_else(|| {
+        let mut services = HashSet::new();
+        services.insert(CONFIGURATION_SERVICE.to_owned());
+        services
+    });
+
+    // Prepare list of service factories from system and user services
+    let mut resulting_factories = Vec::new();
+
+    // Process system services first
+    let mut all_system_service_factories = service_factories();
+    for service_name in system_services {
+        match all_system_service_factories.remove(&service_name) {
```

Comment:
(Again) I find this code unreadable and unnatural, because it removes elements for no apparent reason when it seems to just need to check if a factory is present and get it.

Reply:
Since it wasn't me, who introduced this piece of code, I considered it already reviewed and correct. From my understanding, removing semantic is used here in order to not to put service factory to the builder more than once. If we rely on checking only then it is possible to put it twice and more. @vitvakatu could you please elaborate on this (why removing is used)?
But on the other hand, we have a test for duplicates in the `system_services` section which relies on the `TOML` deserialization logic and thus with this particular implementation duplicates are not possible -> we could switch to checking.

Reply:
Honestly, can't remember. I think `get` is also okay

Reply:
Well, it turned out to be the more clear way to write this piece of code WRT the Rust move semantics with `.remove()` than using `.get()` or whatever else. I added a comment that notes that we're _moving_ factory instance to the new collection.